src/odemis/driver/smaract.py

```diff
@@ -4913,6 +4922,406 @@ def _cancelReference(self, future):
+
+    def _getReadableEvent(self, event):
+        """
+        Get readable event name for logging purposes.
+
+        :param error: (si.EventType or ps.EventType) The event.
+        :return: (str) The readable event name.
+
+        """
+        ev_name = "(0x{:04X})".format(event.type)
+        if event.type in set(ev.value for ev in si.EventType):
+            ev_name = "si.EventType." + si.EventType(event.type).name + " " + ev_name
+        elif event.type in set(ev.value for ev in ps.EventType):
+            ev_name = "ps.EventType." + ps.EventType(event.type).name + " " + ev_name
+        return ev_name
+
+    def _cancelReference(self, future):
+        """
+        Cancels the referencing procedure.
+        future (Future): the future to stop
+        """
+        logging.debug("Cancelling referencing...")
+        future._must_stop.set()  # tell the thread taking care of the referencing it's over
+
+        # Synchronise with the ending of the future
+        with future._moving_lock:
+            si.Cancel(self.id)
+
+        return True
+
+    def _createReferenceFuture(self):
+        """
+        returns: (CancellableFuture)
+        """
+        f = CancellableFuture()
+        f._must_stop = threading.Event()  # cancel of the current future requested
+        f._moving_lock = threading.Lock()  # taken while moving
+        f.task_canceller = self._cancelReference
+        return f
+
+    @isasync
+    def reference(self, _=None):
+        """
+        This is not a "normal" referencing procedure since the hardware doesn't have any actuators. Instead,
+        it performs an internal adjustment routine which is required to get accurate position values.
+        """
+        f = self._createReferenceFuture()
+        self._executor.submitf(f, self._doReference, f)
+        return f
+
+    def _getPosition(self):
+        """
+        Getter for .position VA. Requests position from device.
+        returns: dict (str --> float)
+        """
+        # Polling thread is running all the time except during referencing
+        if self._polling_thread is None:
+            logging.warning("Cannot report position, device needs to be referenced first.")
+            return {}
+        pos = {}
+        for name, num in self._channels.items():
+            pos[name] = si.GetValue_f64(self.id, num, 0)
+        return pos
+
+    def terminate(self):
+        # should be safe to close the device multiple times if terminate is called more than once.
+        if self._polling_thread:
+            self._polling_thread.cancel()
+        if self._executor:
+            self._executor.cancel()
+            si.Close(self.id)
+            self._executor = None
+        super(PicoscalePythonSDK, self).terminate()
+
+    @roattribute
+    def axes(self):
+        """ dict str->Axis: name of each axis available -> their definition."""
+        return self._axes
+
+    def _wait_for_progress_event(self, end_state, timeout=float("inf")):
```

Comment:
Actually, with Python 3 you can write `math.inf`, which is nicer to read. Could you update the code to use this writing?